crates/pico/src/database.rs

```diff
@@ -0,0 +1,173 @@
+                node.time_calculated = time_calculated;
+                node.time_verified = self.current_epoch;
+                time_calculated
+            } else {
+                self.nodes
+                    .get(&node_id)
+                    .expect("node should exist. This is indicative of a bug in Isograph.")
+                    .time_calculated
+            }
+        } else {
+            let (value, dependencies, time_calculated) =
+                self.call_inner_fn_and_collect_dependencies(param_id, inner_fn);
+            self.nodes.put(
+                node_id,
+                DerivedNode {
+                    time_verified: self.current_epoch,
+                    time_calculated,
+                    dependencies,
+                    inner_fn,
+                },
+            );
+            self.values.put(node_id, value);
+            time_calculated
+        };
+
+        self.register_dependency(node_id, time_calculated);
+        node_id
+    }
+
+    fn any_dependency_changed(&mut self, node_id: &NodeId) -> bool {
+        let dependencies = self
+            .nodes
+            .get(node_id)
+            .expect("node should exist. This is indicative of a bug in Isograph.")
+            .dependencies
+            .iter()
+            .filter_map(|dep| {
+                if dep.time_verified_or_calculated != self.current_epoch {
+                    Some((dep.node_to, dep.time_verified_or_calculated))
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+
+        dependencies
+            .into_iter()
+            .any(|(dep_node_id, time_verified)| match dep_node_id.kind {
+                NodeKind::Source => self.source_node_changed(&dep_node_id, time_verified),
+                NodeKind::Derived => self.derived_node_changed(&dep_node_id),
+            })
+    }
+
+    fn source_node_changed(&mut self, node_id: &NodeId, time_verified: u64) -> bool {
+        match self.sources.get(node_id) {
+            Some(source) => source.time_calculated > time_verified,
+            None => true,
+        }
+    }
+
+    fn derived_node_changed(&mut self, node_id: &NodeId) -> bool {
+        if !self.params.contains(&node_id.param_id) {
+            return true;
+        }
+        let new_value = (self.nodes.get(node_id).unwrap().inner_fn)(self, node_id.param_id);
+        match self.values.get(node_id) {
+            Some(value) => *value != new_value,
+            None => true,
+        }
+    }
+
+    fn call_inner_fn_and_collect_dependencies(
+        &mut self,
+        param_id: ParamId,
+        inner_fn: impl Fn(&mut Database, ParamId) -> Box<dyn DynEq>,
+    ) -> (
+        Box<dyn DynEq>,  /* value */
+        Vec<Dependency>, /* dependencies */
+        u64,             /* time_calculated */
+    ) {
+        self.dependency_stack.push(vec![]);
+        let value = inner_fn(self, param_id);
+        let registred_dependencies = self
```

Comment:
typo: `registred`

Also, perhaps we should make a separate function that does something like

```
let (result, dependencies) = with_additional_item_on_dependency_stack(
  || do_something()
);
```

Reply:
Yeah, like it!